### PR TITLE
Add proposal for userns build

### PR DESCRIPTION
**Pull request by chmouel**
Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

### CODE REVIEW

File: enhancements/build-as-userns.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  propogate properly when running from a pipelinerun, the PR is here :
+
+  https://github.com/tektoncd/pipeline/pull/4478
+
+
+### Solution 2
+
+* we need to enforce running as the `build` user on container.
+
+* To allow that we could have a new SCC with less right than the pipelines scc
```
Comment (siamaksade) [inline enhancements/build-as-userns.md:58]: What are the downsides of modifying the existing pipeline sa and reduce its rights?

File: enhancements/build-as-userns.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  and only able to run as the user build instead of the anyuid rights we have
+  now.
+
+* the openshift-pipelines operator then would have to create two different SA and two different SCC in each namespaces.
+
+* we would need to modify the buildah tekton tasks to use that SA.
```
Comment (siamaksade) [inline enhancements/build-as-userns.md:64]: How could we enforce using an SA only with the buildah/s2i tasks?

File: enhancements/build-as-userns.md (hunk 1 above)
Comment (chmouel) [reply enhancements/build-as-userns.md:58]: user by default won't be able to run anything as root? i would love to have this but expect more bugs report from users, maybe we should bit it and run with it  ?

File: enhancements/build-as-userns.md (hunk 2 above)
Comment (chmouel) [reply enhancements/build-as-userns.md:64]: by modifying the tasks we ship ? 

anything else is going to have to do that themselves...

I am working on getting upstream buildah catalog task to run as user by default too..

File: enhancements/build-as-userns.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Build containers images as user
+
+## Goals
+
+There is two different scenarios we want to achieve :
+
+1) Allow user to run buildah as root inside the container but running as user on the host
```
Comment (vdemeester) [inline enhancements/build-as-userns.md:7]: This is user namespace (from the host view point)

File: enhancements/build-as-userns.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+2) Running as user inside the container.
```
Comment (vdemeester) [inline enhancements/build-as-userns.md:9]: This is rootless (similar to what we can achieve on desktop/laptops with `podman`/`buildah`)

File: enhancements/build-as-userns.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+### Scenario 1 Running as root on container as user on host
+
+* Openshift 4.10 will introduces a new "builder" workload on cri-o config. This
+  allows running container as user on the host but shows as root on host. The
```
Comment (vdemeester) [inline enhancements/build-as-userns.md:14]: typo: "running container as user on the host but shows as root on the container"

File: enhancements/build-as-userns.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+* We would have to modify the buildah task to add the annotations into it.
+
+* We would need this PR merged to get the subuids and subgids added on hosts :
+
+  https://github.com/openshift/machine-config-operator/pull/2906
+
+* There is currently a bug on pipeline where the annotation on task doesn't
+  propogate properly when running from a pipelinerun, the PR is here :
+
+  https://github.com/tektoncd/pipeline/pull/4478
```
Comment (vdemeester) [inline enhancements/build-as-userns.md:51]: We can update this, as the PR is merged and will be in 1.7.

File: enhancements/build-as-userns.md (hunk 1 above)
Comment (vdemeester) [reply enhancements/build-as-userns.md:58]: Long term we want to remove `pipeline` SA altogether, but as @chmouel said, as everybody use this by default, we may break users, so we should at least warn them ahead of time (deprecation policy, …).

File: enhancements/build-as-userns.md (hunk 2 above)
Comment (vdemeester) [reply enhancements/build-as-userns.md:64]: ServiceAccount are attached on runtime objects (`TaskRun` and `PipelineRun`) so this wouldn't work by modifying the `Task`. We would need to have something (a webhook, …) that knows when to change the default `ServiceAccount` depending on condition (an annotation, an image, … on the Task or Pipeline).

File: enhancements/build-as-userns.md (hunk 3 above)
Comment (chmouel) [reply enhancements/build-as-userns.md:7]: sounds good! thanks for clarifying the wordings.

**Pull request closed without merging**